### PR TITLE
Merge make dummies

### PR DESCRIPTION
**Pull request by aredelmeier**

### CODE REVIEW

File: R/features.R
Comment (martinju) [inline R/features.R:214]: Suggesting data -> x_train or data_train, newdata -> x_test or data_test. There is nothing new about newdata here (while there actually is in `apply_dummies`.

File: R/features.R
Comment (martinju) [inline R/features.R:228]: For this and similar places: 

We say that data and newdata must be data.table or data.frame, but still we call as.data.frame, so it could have been matrix. Do we need to handle matrix? If not, we could skip the cumbersome as.data.frame before as.data.table here. We could/should then check that there are no other column classes than is.numeric and is.factor. 

If one passes a data.frame with characters, currently this converts the data.frame characters to a factor with the levels present in that data.frame. If the same or not found in the test data, then the error of not identical levels might be confusing as the user have not specified any levels, our function has created it for them. If we require factor from the beginning we avoid this.

File: R/features.R
Comment (martinju) [inline R/features.R:239]: I think we should specifiy what features are missing doing something like 
paste0("The feature(s) ",paste0(features[!(features %in% names(newdata))],collapse = ", ")," are missing from newdata.")

**Review by martinju — COMMENTED** [review]
Have you considered combining the training data and testing data in make_dummies, extract the factors, contrasts etc, and then just calling apply_dummies both for the training and test data? I would imagine that would make the code cleaner, but there might be issues that I am not thinking of. 

If that is not feasible I would change some variable names in make_dummies like is_factor_all, which refers to data, and not both data and newdata, while is_factor_new refers to the new data (which is the testing data). Using the train and test tags instead of data and newdata makes it easier for the user to understand the error messages too (they just passed the training and testing data).

File: R/features.R
Comment (aredelmeier) [reply R/features.R]: Thanks for the suggestion. I have now done this.

File: R/features.R
Comment (aredelmeier) [reply R/features.R:185]: Good point! Yes "data" is needed when using apply_dummies() in shapr()

Comment (martinju) [reply R/features.R:186]: Are you sure? It seems all you use obj$data for within apply_dummies you have already stored in other parts of obj.

File: R/features.R
Comment (aredelmeier) [reply R/features.R]: This is dealt with.

File: R/features.R
Comment (aredelmeier) [reply R/features.R:185]: Good point! Fixed

File: R/features.R
Comment (martinju) [inline R/features.R:220]: I have not tried, but is model.frame <- model.matrix <- NULL really needed? They are just functions...

File: R/features.R
Comment (martinju) [inline R/features.R:241]:
```suggestion
  if (any(duplicated(features))) {
```

File: R/features.R
Comment (martinju) [inline R/features.R:258]:
```suggestion
  # Check if the features all have class "integer", "numeric" or "factor
  if (!all(sapply(traindata, class) %in% c("integer", "numeric", "factor"))) {
    stop("All traindata must have class integer, numeric or factor.")
  }
  if (!all(sapply(testdata, class) %in% c("integer", "numeric", "factor"))) {
    stop("All testdata must have class integer, numeric or factor.")
```

File: R/features.R
Comment (martinju) [inline R/features.R:337]: Same as above. model.frame and model.matrix should not be needed, I think.

File: R/features.R
Comment (martinju) [inline R/features.R:356]:
```suggestion
  if (!all(sapply(testdata, class) %in% c("integer", "numeric", "factor"))) {
    stop("All testdata must have class integer, numeric or factor.")
```

**Review by martinju — COMMENTED** [review]
See the few minor comments. The rest looks good. I will wait for the tests to approve.

File: inst/scripts/example_ctree_method.R
Comment (martinju) [inline inst/scripts/example_ctree_method.R:89]:
```suggestion
# Specifying the phi_0, i.e. the expected prediction without any features
```

File: tests/testthat/test-features.R
Comment (martinju) [inline tests/testthat/test-features.R:315]:
```suggestion
```

**Review by martinju — COMMENTED** [review]
Just tiny changes. Will commit them + some documentation stuff done locally before merging.

**Review by martinju — COMMENTED** [review]
tiny fixes

**Review by martinju — APPROVED** [review]

**Review by aredelmeier — COMMENTED** [review]
This looks fine to me

**Pull request merged**
